src/ray/raylet/wait_manager.h

```diff
@@ -74,7 +74,7 @@ class WaitManager {
     /// The number of required objects.
     const uint64_t num_required_objects;
     /// The objects that have been locally available.
-    std::unordered_set<ObjectID> ready;
+    std::vector<ObjectID> ready;
```

Comment:
Thanks for fixing the issue!

I actually think `set` type is better here: it reflects the fact that for duplicate objects, it's all or nothing semantics. Could we instead fix the `if` conditions so instead of directly comparing `wait_request.ready.size()` and `wait_request.num_required_objects` we calculate how many `wait_request.object_ids` are in the `ready` set? It can be a helper function in WaitRequest.

Reply:
How about we dedup the object ids before calling `WaitManager::Wait` and add a check at the beginning of it to disallow duplicate object ids? I noticed that we don't allow `ray.wait` on duplicate ObjectRefs, so I guess we only need to dedup actor task dependencies.

Reply:
That sgtm.

Reply:
Can you check that will work with nested ObjectRefs? Like this:


```python
f.remote(arg, arg)  # Not allowed.
f.remote(arg, [arg])  # Allowed.
```

I'm wondering about this use case because I think this can actually come up quite easily if the application is passing ObjectRefs around in arbitrary structures. Conversely, it's not a problem in ray.wait() since we require the user to pass all ObjectRefs in a list.

Reply:
@stephanie-wang I don't think we need to handle this specially because `[obj]` won't be unboxed automatically when executing a task, so the worker doesn't need to wait until `obj` is local.

Reply:
Got it, thanks!

Reply:
Reverted logic in WaitManager and added duplication check.